### PR TITLE
Ledger Signup flow

### PR DESCRIPTION
**Pull request by danimoh**

### CODE REVIEW

File: src/views/SignupSuccess.vue [hunk 1]
```diff
@@ -23,31 +26,54 @@
...
-    @Getter private hasWallets!: boolean;
-
-    private walletLabel: string = 'Keyguard Wallet';
-    private accountLabel: string = 'Standard Account';
+    private static readonly STEPS_KEYGUARD_SIGNUP = 6;
+    private static readonly STEPS_LEDGER_SIGNUP = 3;
+    private static readonly DEFAULT_KEYGUARD_WALLET_LABEL = 'Keyguard Wallet';
+    private static readonly DEFAULT_LEDGER_WALLET_LABEL = 'Ledger Wallet';
+    private static readonly DEFAULT_KEYGUARD_ACCOUNT_LABEL = 'Standart Account';
+    private static readonly DEFAULT_LEDGER_ACCOUNT_LABEL = 'Ledger Account';
```
Comment (sisou) [inline src/views/SignupSuccess.vue]: Should go into `Constants.ts`.

File: src/components/IdenticonSelector.vue [hunk 2]
```diff
@@ -0,0 +1,233 @@
+<template>
+    <div class="identicon-selector" :class="{ selected: !!selectedAccount }">
+        <h1 class="nq-h1">Choose an Avatar</h1>
+        <div class="identicons">
+            <div class="center" v-if="displayedAccounts.length === 0">
+                <div class="loading-animation"></div>
+                <h2>Mixing colors</h2>
+            </div>
+            <div class="wrapper" @click="_selectAccount(account)"
+                 :class="{ selected: selectedAccount === account }" v-for="account in displayedAccounts">
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/components/IdenticonSelector.vue]: `:key` is not required. If a `:key` is provided, it uses the rendered dom elements only for the data entry with this key. If no key is provided, vue tries to reuse the rendered dom elements.

File: src/components/IdenticonSelector.vue [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    import { Component, Prop, Watch, Vue } from 'vue-property-decorator';
+    import { Identicon } from '@nimiq/vue-components';
+    import { AccountInfo } from '../lib/AccountInfo';
+
+    @Component({components: { Identicon }})
+    class IdenticonSelector extends Vue {
+        private static readonly IDENTICONS_PER_PAGE = 7;
+
+        @Prop({ default: () => [], type: Array })
+        public accounts!: AccountInfo[];
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/components/IdenticonSelector.vue]: From [Vue Documentation](https://vuejs.org/v2/guide/components-props.html):

> Object or array defaults must be returned from a factory function.

File: src/components/LedgerUi.vue [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+                this._showInstructions(null);
+                break;
+            case LedgerApi.StateType.LOADING:
+                this._onStateLoading();
+                break;
+            case LedgerApi.StateType.REQUEST_PROCESSING:
+                this._onRequest(state);
+                break;
+            case LedgerApi.StateType.REQUEST_CANCELLING:
+                this._showInstructions('', 'Please cancel the call on your Ledger');
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/components/LedgerUi.vue]: Changed it to `request`.

File: src/components/LedgerUi.vue [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+                this._onStateLoading(); // show as still loading / retrying
+                break;
+            case LedgerApi.ErrorType.NO_BROWSER_SUPPORT:
+                this._showInstructions('', 'Ledger not supported by browser or support not enabled.');
+                break;
+            case LedgerApi.ErrorType.APP_OUTDATED:
+                this._showInstructions('', 'Your Nimiq App is outdated. Please update using Ledger Live.');
+                break;
+            case LedgerApi.ErrorType.WRONG_LEDGER:
+                this._showInstructions('', 'The connected Ledger is not the correct one.');
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/components/LedgerUi.vue]: Changed it to `The connected Ledger is not the one this account belongs to.`

File: src/components/LedgerUi.vue [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+        display: flex;
+    }
+
+    @keyframes ledger-connect-cable {
+        0% {
+            transform: translateX(-50%);
+        }
+        25% {
+            transform: translateX(0);
+        }
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/components/LedgerUi.vue]: Yes, but that's less readable

File: src/components/WalletIdentifier.vue [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+</template>
+
+<script lang="ts">
+import { Component, Prop, Watch, Vue } from 'vue-property-decorator';
+import { Identicon } from '@nimiq/vue-components';
+import { AccountInfo } from '../lib/AccountInfo';
+
+@Component({components: { Identicon }})
+export default class WalletIdentifier extends Vue {
+    @Prop({ default: () => [], type: Array })
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/components/WalletIdentifier.vue]: See above

File: src/components/WalletIdentifier.vue [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        height: 100%;
+        transform: scale(1.15); /* to avoid that identicon appears smaller due to white border */
+    }
+
+    .wallet-identifier .placeholder {
+        background: url('data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 58 49"><path fill="%231F2348" d="M56.8 22.2L45 2.7a4.8 4.8 0 0 0-4-2.3H17c-1.7 0-3.2.9-4 2.3L1 22.2a4.3 4.3 0 0 0 0 4.6L13 46.3a4.8 4.8 0 0 0 4.1 2.3h23.7c1.7 0 3.3-.9 4.1-2.3l11.9-19.5c.9-1.4.9-3.2 0-4.6z" opacity=".08"/></svg>') no-repeat;
+    }
+
+    .wallet-identifier .pop-in {
+        animation: pop-in 0.5s cubic-bezier(.99,1.54,.49,1.15) backwards;
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/components/WalletIdentifier.vue]: I don't know. I took it from @nibhar 's PR https://github.com/nimiq/keyguard-next/pull/151.
However, I just saw, that he changed that easing function again and updated it here too.

File: src/router.ts [hunk 9]
```diff
@@ -16,13 +16,16 @@ const SignupTypeSelector      = () => import(/*webpackChunkName: "signup"*/ './v
...
+
 const Login                   = () => import(/*webpackChunkName: "login"*/ './views/Login.vue');
 const LoginSuccess            = () => import(/*webpackChunkName: "login"*/ './views/LoginSuccess.vue');
 const LoginErrorHandler       = () => import(/*webpackChunkName: "login"*/ './views/LoginErrorHandler.vue');
 
 const Export                  = () => import(/*webpackChunkName: "export"*/ './views/Export.vue');
 
-const ChangePassphrase        = () => import(/*webpackChunkName: "change-passphrase"*/ './views/ChangePassphrase.vue');
+const ChangePassphrase        = () => import(/*webpackChunkName: "change-passphrase"*/
+    './views/ChangePassphrase.vue');
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/router.ts]: Hum, probably tried changing the layout or renaming something and then forgot to undo it.

File: src/router.ts [hunk 10]
```diff
@@ -148,6 +151,11 @@ export default new Router({
             component: SignupTypeSelector,
             name: `${RequestType.SIGNUP}`,
         },
+        {
+            path: `/${RequestType.SIGNUP}/ledger`,
+            component: SignupLedger,
+            name: `${RequestType.SIGNUP}-ledger`,
+        },
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/router.ts]: Why would the `/signup/ledger` route not work anymore?

File: src/components/IdenticonSelector.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (sisou) [reply src/components/IdenticonSelector.vue]: Fascinating! How did you find out about this?

File: src/components/LedgerUi.vue (hunk 6 above)
[the remark this answers is not in the document]
Comment (sisou) [reply src/components/LedgerUi.vue]: I think it is, if it only contains one rule :P

[the remark this answers is not in the document]
Comment (sisou) [reply src/components/LedgerUi.vue]: It's minified during build anyway. It's fine.

File: src/lib/WalletInfoCollector.ts [hunk 11]
```diff
@@ -41,14 +46,17 @@ export default class WalletInfoCollector {
...
+            WalletInfoCollector._addAccounts(walletInfo, initialAccounts, undefined);
+            // fetch balances and update again
+            initialAccountsPromise = WalletInfoCollector._getBalances(initialAccounts).then(async (balances) => {
+                WalletInfoCollector._addAccounts(walletInfo, initialAccounts, balances);
+                if (onUpdate) onUpdate(walletInfo, await derivedAccountsPromise);
+            });
         } else {
             initialAccountsPromise = Promise.resolve();
         }
+        if (onUpdate) onUpdate(walletInfo, await derivedAccountsPromise);
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/WalletInfoCollector.ts]: That also works :D

File: src/router.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (sisou) [reply src/router.ts]: It works for the state `master` is in during this PR, but in a following PR (about signup restructuring) this changes. I can do it in that PR then, it's fine.

File: src/views/SignupLedger.vue [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+                : 'Welcome to the Nimiq Blockchain.';
+    }
+
+    private get loaderStatus() {
+        if (this.state !== SignupLedger.State.FETCHING_ACCOUNTS) return '';
+        else if (this.failedFetchingAccounts) return 'Failed to fetch accounts. Retrying...';
+        else {
+            const count = !this.walletInfo ? 0 : this.walletInfo.accounts.size;
+            return count > 0
+                ? `Imported ${count} account${count > 1 ? 's' : ''} so far.`
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/views/SignupLedger.vue]: No, as you can see in the line above :D

File: src/views/SignupLedger.vue [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+            ));
+        }
+        this.$forceUpdate(); // because vue does not recognize changes in waletInfo.accounts map // TODO verify
+    }
+
+    @Emit()
+    private async done() {
+        this.state = SignupLedger.State.FINISHED;
+        setTimeout(() => {
+            const result: SignupResult = {
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/views/SignupLedger.vue]: I prefer creating the result from the most up to date data just before resolving the call.

File: src/views/SignupLedger.vue (hunk 12 above)
[the remark this answers is not in the document]
Comment (sisou) [reply src/views/SignupLedger.vue]: Ha, damn! :laughing:

File: src/components/IdenticonSelector.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (danimoh) [reply src/components/IdenticonSelector.vue]: By reading the documentation ;)
Also Polymer actually had a similar restriction for default data.
The reason is that for non primitive types the same data instance would be shared across all component instances, therefore one can provide a factory method to create a new data instance for every component instance.

File: src/components/WalletIdentifier.vue (hunk 8 above)
[the remark this answers is not in the document]
Comment (danimoh) [reply src/components/WalletIdentifier.vue]: BTW, I asked him about this easing function and he told me that Julian designed it this way to create a bounce back effect

**Pull request merged**